Fix IndexError: list index out of range in resampler

llucax opened the pull request:
When awaiting the resamplers to finish resampling, we need to make a copy of the resamplers dict, because new resamplers could be added or removed while we are awaiting, which would cause the results to be out of sync.

Here is a traceback of an instance of this bug:

```
2024-11-25 12:45:26,052,52 WARNING  [_broadcast.py:416] Broadcast receiver [Broadcast:GrpcStreamBroadcaster-raw-component-data-2006:_Receiver] is full. Oldest message was dropped.
2024-11-25 12:45:26,052,52 ERROR    [_resampling.py:195] The resample() function got an unexpected error, restarting...
Traceback (most recent call last):
  File "/home/marenz/frequenz/sdk/src/frequenz/sdk/microgrid/_resampling.py", line 179, in _log_resampling_task_error
    resampling_task.result()
  File "/home/marenz/frequenz/sdk/src/frequenz/sdk/timeseries/_resampling.py", line 509, in resample
    {
  File "/home/marenz/frequenz/sdk/src/frequenz/sdk/timeseries/_resampling.py", line 514, in <dictcomp>
    if isinstance(results[i], (Exception, asyncio.CancelledError))
                  ~~~~~~~^^^
IndexError: list index out of range
```

llucax committed:
Clear release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Fix out of bounds error in the resampler

When awaiting the resamplers to finish resampling, we need to make a
copy of the resamplers dict, because new resamplers could be added or
removed while we are awaiting, which would cause the results to be
out of sync.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -495,8 +495,13 @@ async def resample(self, *, one_shot: bool = False) -> None:
+            # current resamplers, and since we await here, new resamplers could be added
+            # or removed from the dict while we awaiting the resampling, which would
+            # cause the results to be out of sync.
+            resamplers = self._resamplers.copy()
```
Copying keys should be enough:
`resampler_keys = [self_.resamplers.keys()]`

llucax replied:
No, we actually use the value (`source`) too, to set in which resampler the error was found.

ela-kotulska-frequenz replied:
But : 
* `self._resamplers: dict[Source, _StreamingHelper] = {}` - so you will copy source.
* and `enumerate(resamplers)` - enumerates over keys.

llucax replied:
Right, right, right. I got confused because we are using the `values()` right after the copy, but when we access those `values()` there was no `await` yet, so we can use the original `dict` there. :+1:

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -508,7 +513,7 @@ async def resample(self, *, one_shot: bool = False) -> None:
                 dict[Source, Exception | asyncio.CancelledError],
                 {
                     source: results[i]
-                    for i, source in enumerate(self._resamplers)
+                    for i, source in enumerate(resamplers)
                     # CancelledError inherits from BaseException, but we don't want
                     # to catch *all* BaseExceptions here.
                     if isinstance(results[i], (Exception, asyncio.CancelledError))
```
Not related, but it looks like perfect case for using zip :) 
`{source: result for source, result in zip(resamplers_keys, results) if isinstance(result, (Exception, asyncio.CancelledError))`

llucax replied:
Yeah, I could try if it helps mypy do the type-narrowing, but I doubt it.

llucax replied:
It worked, not sure if maybe it also worked now without the `zip()` too with newer `mypy` versions or not, but nice that we don't need the cast anymore!

llucax committed:
Use zip() to simplify the matching of sources and results

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove the unnecessary cast for exceptions in remove_timeseries

Either `mypy` was improved in general or using `zip()` helped it to
figure out the type narrowing, but it seems it can now narrow the type
of `exceptions` without the need of a cast.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Updated to:
* Copy only the keys
* Use `zip()`
* Remove the unnecessary cast

ela-kotulska-frequenz reviewed (approved):
Thanks!
